lms/djangoapps/discussion/tasks.py

```diff
@@ -32,6 +33,30 @@
 ROUTING_KEY = getattr(settings, 'ACE_ROUTING_KEY', None)
 
 
+@task(base=LoggedTask)
+def update_discussions_map(context):
+    course_key = CourseKey.from_string(context['course_id'])
+    discussions_id_map = {}
+
+    with modulestore().bulk_operations(course_key):
+        course = modulestore().get_course(course_key, depth=None)
+        blocks_stack = [course]
+        while blocks_stack:
+            curr_block = blocks_stack.pop()
+            children = curr_block.get_children() if curr_block.has_children else []
+            if (
+                curr_block.category == 'discussion' and
+                hasattr(curr_block, 'discussion_id') and
+                curr_block.discussion_id
+            ):
+                discussions_id_map[curr_block.discussion_id] = unicode(curr_block.scope_ids.usage_id)
+
+            # Add this blocks children to the stack so that we can traverse them as well.
+            blocks_stack.extend(children)
```

Comment:
Is this method of traversing the course structure to find discussion blocks more efficient than using `get_items` with the `category` kwarg?

Reply:
hmm.. not sure.  Perhaps @ormsbee knows for sure.
Calling with `depth=None` gets the entire structure and then we're instantiating each xBlock, which is definitely costly.
Calling `get_items` would limit the number of xBlocks that we instantiate, but it may still traverse the entire structure on-disk.  (Or may not depending on whether the modulestore is smart enough to not open each block's definition when filtering by `category`).
Alternatively, we use a downstream signal from `BlockStructure` and iterate through all the blocks from `BlockStructure` instead of directly from `Modulestore`.

Reply:
I suspect `get_items` would be a little faster, but I wouldn't know for sure without profiling. I'd say just do the simpler one to start off.